### PR TITLE
Song list page

### PR DESCRIPTION
**Pull request by stefanikoetsier**
This branch implements the song list page which can be found at:
http://127.0.0.1:8000/song-list

The page shows the four following columns:

- Artist, 
- Title, 
- Average rating
- Number of ratings.

A new database table for ratings is created. And unit tests for the new view are included.

### CODE REVIEW

File: song_rater/models/song.py [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=['title', 'artist'], name='unique_song')
+        ]
+
+
+class Rating(models.Model):
+    song = models.ForeignKey(Song, on_delete=models.CASCADE, related_name='ratings')
+    rating = models.PositiveSmallIntegerField(validators=[MaxValueValidator(5)])
```
Comment (SebastiaanZ) [inline song_rater/models/song.py]: Is `0` or `1` supposed to be your minimum rating value?

File: song_rater/templates/song_rater/song_list.html [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        <td>{{ song.artist }}</td>
+        <td>{{ song.title }}</td>
+        <td>{{ song.avg_rating }}</td>
+        <td>{{ song.n_ratings }}</td>
+    </tr>
+    {% endfor %}
+
+    </table>
+</body>
+</html>
```
Comment (SebastiaanZ) [inline song_rater/templates/song_rater/song_list.html]: This file should probably get a newline at the end.

File: song_rater/templates/song_rater/song_list.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    </tr>
+
+    {% for song in song_table %}
+    <tr>
+        <td>{{ song.artist }}</td>
+        <td>{{ song.title }}</td>
+        <td>{{ song.avg_rating }}</td>
+        <td>{{ song.n_ratings }}</td>
+    </tr>
+    {% endfor %}
```
Comment (SebastiaanZ) [inline song_rater/templates/song_rater/song_list.html]: You could make this more dynamic, as in displaying a message when no songs have been added to the site yet. It's not an occurrence that will happen a lot after the site goes into production, but it may be a good opportunity to look into template tags like `for empty endfor`.

If you want, you could even only print the header row if there are songs available by printing them in the first iteration of the template-for loop (there's another tag for that).

File: song_rater/tests.py [hunk 4]
```diff
@@ -29,3 +29,16 @@ def test_add_song_post(self):
...
+    def setUp(self):
+        song = Song.objects.create(artist='Apple', title='Pie')
+        song.ratings.create(rating=3)
+
+    def test_song_list(self):
+        response = self.client.get('/song-list')
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, 'Apple', count=1)
+        self.assertContains(response, 'Pie', count=1)
+        self.assertContains(response, '3', count=1)
```
Comment (SebastiaanZ) [inline song_rater/tests.py]: Would it be interesting to test another boundary condition: How your song listing works if there are *no* elements in the song table yet?

It may also be a good idea to test a listing with multiple songs: Does that work? (I see no reason for why it wouldn't work, but it tests the situation in which returning one song works, but the listing fails with multiple items.)

File: song_rater/views.py [hunk 5]
```diff
@@ -13,5 +15,30 @@ class SongCreate(CreateView):
...
 
+class SongList(ListView):
+    model = Song
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        song_table = []
+
+        for song in Song.objects.all().prefetch_related('ratings'):
```
Comment (SebastiaanZ) [inline song_rater/views.py:26]: not really a comment about this section of code, but I think you'll like [this article](https://docs.djangoproject.com/en/3.1/topics/db/optimization/) in the django docs.

File: song_rater/models/song.py (hunk 1 above)
Comment (SebastiaanZ) [reply song_rater/models/song.py]: Your commit message explains that `0` is the intended minimum, which is exactly what this does.

File: song_rater/models/song.py [hunk 6]
```diff
@@ -6,6 +6,11 @@ class Song(models.Model):
     title = models.CharField(max_length=100)
     artist = models.CharField(max_length=100)
 
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=['title', 'artist'], name='unique_song')
+        ]
+
```
Comment (SebastiaanZ) [inline song_rater/models/song.py:13]: Do you want to add a test for this constraint? Just to make sure you handle a failure of the restraint in the intended way? (What is the intended way? A nice message to the user or a generic non-201 error page?)

**Review by SebastiaanZ — CHANGES_REQUESTED** [review]
I'd like to see a few more tests that test your listing in a few more boundary conditions:

- 0 songs available
- multiple songs available

A creation test would also not be a bad idea: You should be able to use the client and a `post` action to test what happens when you post a new song. This also allows you to test if your unique constraint properly triggers (and if you handle that in the way you like). This could be a bit more difficult, though, so don't worry if you don't have the time for it.

File: song_rater/views.py [hunk 7]
```diff
@@ -13,5 +15,30 @@ class SongCreate(CreateView):
...
+            ratings = song.ratings.all()
+            n_ratings = ratings.count()
+            avg_rating = ratings.aggregate(Avg('rating'))['rating__avg'] or '-'
+
+            song_table.append({
+                'artist': song.artist,
+                'title': song.title,
+                'avg_rating': avg_rating,
+                'n_ratings': n_ratings,
```
Comment (SanderBeekhuis) [inline song_rater/views.py]: I think you can skip the introduction of the `n_ratings` variable.  `ratings.count()` is just as clear.  
Just stick the expresion you evaluate directly into the dict. :) 

The same might be true for 'avg_rating'

File: song_rater/views.py [hunk 8]
```diff
@@ -13,5 +15,30 @@ class SongCreate(CreateView):
     success_url = '/song-added'
 
 
+class SongList(ListView):
+    model = Song
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        song_table = []
```
Comment (SanderBeekhuis) [inline song_rater/views.py]: I would call this `song_list` or just `songs` but it's a bit subjective.  This kinda implies, to me, that were dealing with some kind of  `Table` object, while this is just an ordinary list.

File: song_rater/tests.py [hunk 9]
```diff
@@ -29,3 +29,16 @@ def test_add_song_post(self):
...
+
+
+class SongListTestCase(TestCase):
+    def setUp(self):
+        song = Song.objects.create(artist='Apple', title='Pie')
+        song.ratings.create(rating=3)
+
+    def test_song_list(self):
+        response = self.client.get('/song-list')
+        self.assertEqual(response.status_code, 200)
```
Comment (SanderBeekhuis) [inline song_rater/tests.py]: 'assertContains' also validates the status code, so this line is not needed.   But it might be that "explicit is better then implicit" [Zen of Python, PEP20] 

https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.SimpleTestCase.assertContains

File: song_rater/tests.py (hunk 4 above)
Comment (SanderBeekhuis) [inline song_rater/tests.py]: What are  you testing when you look in the page for the number `3` ? 

I presume you are testing for an average rating. Maybe state so in a small comment.

**Review by SanderBeekhuis — COMMENTED** [review]
Just a few little nit-picks.  Do with them as you will. 

I don't need to review this again. I trust you and Sebastiaan.

File: song_rater/models/song.py (hunk 1 above)
Comment (stefanikoetsier) [reply song_rater/models/song.py]: I added a minimum value validator to make 1 the minimum value as in the story

File: song_rater/views.py (hunk 7 above)
Comment (stefanikoetsier) [reply song_rater/views.py]: Good find! Thank you

File: song_rater/views.py (hunk 8 above)
Comment (stefanikoetsier) [reply song_rater/views.py]: Agreed, I changed it.

File: song_rater/views.py (hunk 5 above)
Comment (stefanikoetsier) [reply song_rater/views.py:26]: Thank you for sharing. Indeed interesting.

File: song_rater/templates/song_rater/song_list.html (hunk 3 above)
Comment (stefanikoetsier) [reply song_rater/templates/song_rater/song_list.html]: I added the `empty` template tag to show a message to the user. I also found and now use the `forloop.first ` variable which I assume you referred to.

File: song_rater/tests.py (hunk 4 above)
Comment (stefanikoetsier) [reply song_rater/tests.py]: I added tests for both of these cases. Thank you

File: song_rater/models/song.py (hunk 6 above)
Comment (stefanikoetsier) [reply song_rater/models/song.py:13]: I added this constraint test. Django should generate an error message to the user when attempting to add an already existing song. So now I test that that is included in the response.

File: song_rater/tests.py (hunk 9 above)
Comment (stefanikoetsier) [reply song_rater/tests.py]: Thank you for bringing this up. I removed the redundant line of code.

File: song_rater/tests.py (hunk 4 above)
Comment (stefanikoetsier) [reply song_rater/tests.py]: Thank you for finding this, I added a comment for clarification. No need for magic numbers

File: song_rater/models/song.py (hunk 6 above)
Comment (stefanikoetsier) [reply song_rater/models/song.py:13]: I stand corrected. I check for the 302 status code, which only happens when a duplicate song is added

**Comment by stefanikoetsier** [comment]
@SanderBeekhuis @SebastiaanZ Thank you for all the remarks. My code has been updated.

File: song_rater/templates/song_rater/song_list.html (hunk 3 above)
Comment (SebastiaanZ) [reply song_rater/templates/song_rater/song_list.html]: Looks good!

**Review by SebastiaanZ — APPROVED** [review]
